### DISCUSSION TITLE
Implement `async` / `await`

### DISCUSSION THREAD

**Discussion by Razican** [boa-dev/boa#1530] [Ideas]
**ECMASCript feature**
We currently have no support for `async` / `await` in Boa. These two keywords are used to create "promises" that will be executed asynchronously, without blocking the execution of the code that follows.

There are two types of `async` constructs:
 - Async Function Definitions: https://tc39.es/ecma262/#sec-async-function-definitions
 - Async Arrow Function Definitions: https://tc39.es/ecma262/#sec-async-arrow-function-definitions

Implementing this has to be done in two phases. First we need to be able to parse these `async` constructs, and for that, we will need new AST nodes and new parsers, that will be called when finding the `async` keyword.

After the parsing is done, we need to implement the execution. This will be a bit tricky, since we currently just execute the nodes one by one. This means we need to follow the spec in the "Runtime Semantics" for these two constructs, and implement the needed methods.

This might require further talk in #367.

**Comment by Lan2u** [comment]
I'll have a go at this - honestly I've always been very curious as to how the more exotic async stuff is actually implemented so this should be interesting.

**Comment by Lan2u** [comment]
Once this is implemented we will also need to setup benchmark(s) for async operations

**Comment by jasonwilliams** [comment]
> I'll have a go at this - honestly I've always been very curious as to how the more exotic async stuff is actually implemented so this should be interesting.

Most of the asynchronous stuff happens at runtime level rather than engine level. V8 doesn’t handle the event loop, Node does. I think after parse we’ll need to look at how it integrates in other engines. JS in its pure form doesn’t have anything that’s async, setTimeout is injected in for e.g

**Comment by Lan2u** [comment]
> > I'll have a go at this - honestly I've always been very curious as to how the more exotic async stuff is actually implemented so this should be interesting.
> 
> Most of the asynchronous stuff happens at runtime level rather than engine level. V8 doesn’t handle the event loop, Node does. I think after parse we’ll need to look at how it integrates in other engines. JS in its pure form doesn’t have anything that’s async, setTimeout is injected in for e.g

Interesting - I will need to look into this - what is the scope of boa, are we going to implement an event-loop system?

**Comment by Razican** [comment]
I think we could provide a default and very limited `boa_event_loop` crate, just with the minimum to comply with the spec, and to be able to use it in tests. We would add a trait in the `boa` crate with the needed methods, and then rely on the implementers to comply with it.

BTW, any news with the parsing, @Lan2u?

**Comment by Lan2u** [comment]
> I think we could provide a default and very limited `boa_event_loop` crate, just with the minimum to comply with the spec, and to be able to use it in tests. We would add a trait in the `boa` crate with the needed methods, and then rely on the implementers to comply with it.
> 
> BTW, any news with the parsing, @Lan2u?

Yeah, I think that's a good move so we can get stuff tested. Parsing wise I'll try and get the branch ready for review/merge by the end of the week so we can get started on an event_loop (which should be a separate issue I think)

**Comment by Razican** [comment]
I have created a project to keep track of this: https://github.com/boa-dev/boa/projects/3

**Comment by jedel1043** [comment]
Closing, already implemented.

**Reply by mlafeldt** [reply]
Hey there, I noticed that Boa 0.17 now supports promises, which is awesome. 👍

However, using `await` results in a syntax error. Is there some issue that tracks support for `async/await`?

**Reply by mlafeldt** [reply]
Actually, that's not true. This works:

```
❯ boa
>> async function f3() {
  const y = await 20;
  console.log(y); // 20

  const obj = {};
  console.log((await obj) === obj); // true
}

f3();
Promise { <pending> }
20
true
>>
undefined
```

I guess what doesn't work is top-level await.

```
❯ boa -m
>> await 1
Uncaught: SyntaxError: expected token ';', got '1' in expression statement at line 1, col 7
>>
```

**Reply by jedel1043** [reply]
I think that's a "bug" in our cli; at https://github.com/boa-dev/boa/blob/main/boa_cli/src/main.rs#L448-L456 we treat all REPL input as scripts, and only files are treated as modules.

However, I quoted "bug" because it isn't clear to me if the REPL should treat the input as a proper module when the module flag is active.

**Reply by mlafeldt** [reply]
Ah, right. This works. 👍

```
❯ boa <(echo 'console.log(await "yeah")')
Uncaught SyntaxError: expected one of ',' or '}', got 'yeah' in argument list at line 1, col 19

❯ boa -m <(echo 'console.log(await "yeah")')
yeah
```

Thanks.
